### PR TITLE
chore: refactor using-router example

### PR DESCRIPTION
**Pull request by HaNdTriX**
## Changes

- Updated dependencies
- Migrated to Typescript
- Added additional router use cases

## Documentation / Examples

- [x] Make sure the linting passes by running `pnpm lint`
- [x] The "examples guidelines" are followed from [our contributing doc](https://github.com/vercel/next.js/blob/canary/contributing/examples/adding-examples.md)

### CODE REVIEW

File: examples/using-router/components/CustomLink.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+type CustomLinkProps = {
+  children: ReactNode
+  href: string
+  prefetch?: boolean
+  replace?: boolean
+  shallow?: boolean
+} & HTMLAttributes<HTMLAnchorElement>
+
+// typically you want to use `next/link` for this usecase
+// but this example shows how you can also access the router
```
Comment (balazsorban44) [inline examples/using-router/components/CustomLink.tsx:13]: I don't think we need to showcase a way to recreate `next/link`. :thinking:

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Left a small comment.

File: examples/using-router/components/CustomLink.tsx (hunk 1 above)
Comment (HaNdTriX) [reply examples/using-router/components/CustomLink.tsx:13]: oh actually I've just migrated this comment from here:

https://github.com/vercel/next.js/blob/1264196897e11a0561ff9ca9df10c25fb5ee718e/examples/using-router/components/Header.js#L12-L33

Do you have alternative suggestions to show some examples of this imperative api?

Comment (balazsorban44) [reply examples/using-router/components/CustomLink.tsx:13]: I see. In that case, let's leave it then for now. :+1: Thanks for clarifying!

**Review by balazsorban44 — APPROVED** [review]

**Review by ijjk — APPROVED** [review]

**Pull request merged**